data/appRateLimiter.go

```diff
@@ -0,0 +1,147 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package data
+
+import (
+	"encoding/binary"
+	"sync/atomic"
+	"time"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/protocol"
+	"golang.org/x/crypto/blake2b"
+
+	"github.com/algorand/go-deadlock"
+)
+
+// appRateLimiter implements a sliding window counter rate limiter for applications
+type appRateLimiter struct {
+	maxSize           uint64
+	serviceRate       uint64
+	serviceRateWindow time.Duration
+
+	// TODO: consider some kind of concurrent map
+	// TODO: add expiration strategy
+	mu   deadlock.RWMutex
+	apps map[crypto.Digest]*appRateLimiterEntry
+}
+
+func MakeAppRateLimiter(maxSize uint64, serviceRate uint64, serviceRateWindow time.Duration) *appRateLimiter {
+	return &appRateLimiter{
+		maxSize:           maxSize,
+		serviceRate:       serviceRate,
+		serviceRateWindow: serviceRateWindow,
+		apps:              map[crypto.Digest]*appRateLimiterEntry{},
+	}
+}
+
+type appRateLimiterEntry struct {
+	prev     atomic.Uint64
+	cur      atomic.Uint64
+	interval int64 // numeric representation of the current interval value
+}
+
+func (r *appRateLimiter) entry(key crypto.Digest) (*appRateLimiterEntry, bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	entry, ok := r.apps[key]
+	if !ok {
+		entry = &appRateLimiterEntry{}
+		r.apps[key] = entry
+	}
+	return entry, ok
+}
+
+// interval calculates the interval numeric representation based on the given time
+func (r *appRateLimiter) interval(now time.Time) int64 {
+	return now.UnixNano() / int64(r.serviceRateWindow)
+}
+
+// fraction calculates the fraction of the interval that is elapsed since the given time
+func (r *appRateLimiter) fraction(now time.Time) float64 {
+	return float64(now.UnixNano()%int64(r.serviceRateWindow)) / float64(r.serviceRateWindow)
+}
+
+// shouldDrop returns true if the given transaction group should be dropped based on the
+// on the rate for the applications in the group: the entire group is dropped if a single application
+// exceeds the rate.
+func (r *appRateLimiter) shouldDrop(txgroup []transactions.SignedTxn, origin []byte) bool {
+	return r.shouldDropInner(txgroup, origin, time.Now())
+}
+
+func (r *appRateLimiter) shouldDropInner(txgroup []transactions.SignedTxn, origin []byte, now time.Time) bool {
+	var keys []crypto.Digest
+	// TODO: check memory allocs
+	var buf [8 + 4]byte // uint64 + [4]byte
+	for i := range txgroup {
+		if txgroup[i].Txn.Type == protocol.ApplicationCallTx {
+			binary.LittleEndian.PutUint64(buf[:8], uint64(txgroup[i].Txn.ApplicationID))
+			toBeHashed := append(buf[8:], origin...)
+			d := crypto.Digest(blake2b.Sum256(toBeHashed))
```

Comment:
```suggestion
			copy (buf[8:], origin)
			d := crypto.Digest(blake2b.Sum256(buf))
```

Reply:
Was not sure what how to get conn data from p2p so went with safe append option. Eventually it will depends on a underlying cache/map choice, if no bucketing then no need to hash, appid + ip concatenation might be good enough. Also not sure if there should be two levels: appid and ip to give more fairness to other apps.

Reply:
Aren't you trying to make a key that is the hash of the 8 byte app id + the 4 IP bytes?
in your code the length of toBeHashed is 8.
toBeHashed := append(buf[8:], origin...)
buf[8:] is length 4, and origin is length 4.